okhttp-server/src/main/java/okhttp3/webserver/OkHttpServer.java

```diff
@@ -0,0 +1,772 @@
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * A modern HTTP web server.
+ */
+public final class OkHttpServer implements Closeable {
+  private static final Logger logger = Logger.getLogger(OkHttpServer.class.getName());
+
+  private final Set<Socket> openClientSockets =
+      Collections.newSetFromMap(new ConcurrentHashMap<Socket, Boolean>());
+  private final Set<Http2Connection> openConnections =
+      Collections.newSetFromMap(new ConcurrentHashMap<Http2Connection, Boolean>());
+  private long bodyLimit = Long.MAX_VALUE;
+  private ServerSocketFactory serverSocketFactory = ServerSocketFactory.getDefault();
+  private ServerSocket serverSocket;
+  private SSLSocketFactory sslSocketFactory;
+  private ExecutorService executor;
+  private boolean tunnelProxy;
+  private Dispatcher dispatcher = new Dispatcher() {
+    @Override
+    public ServerResponse dispatch(ClientRequest clientRequest) throws InterruptedException {
+      return new ServerResponse();
+    }
+  };
+
+  private int port = -1;
+  private InetSocketAddress inetSocketAddress;
+  private boolean protocolNegotiationEnabled = true;
+  private List<Protocol> protocols = Util.immutableList(Protocol.HTTP_2, Protocol.HTTP_1_1);
+
+  private boolean started;
+
+  private synchronized void maybeStart() {
+    if (started) return;
+    try {
+      start();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public int getPort() {
+    maybeStart();
+    return port;
+  }
+
+  public String getHostName() {
+    maybeStart();
+    return inetSocketAddress.getHostName();
+  }
+
+  public Proxy toProxyAddress() {
+    maybeStart();
+    InetSocketAddress address = new InetSocketAddress(inetSocketAddress.getAddress(), getPort());
+    return new Proxy(Proxy.Type.HTTP, address);
+  }
+
+  @Override
+  public String toString() {
+    return "OkHttpServer[" + port + "]";
+  }
+
+  @Override
+  public void close() throws IOException {
+    shutdown();
+  }
+
+  /**
+   * Sets the dispatcher used to match incoming requests to handlers.
+   */
+  public void setDispatcher(Dispatcher dispatcher) {
```

Comment:
Why not add a `Builder` for such options? Immutability ftw! Please don't repeat design issues of OkHttp pre v3